secure.go

```diff
@@ -101,6 +128,11 @@ func (s *Secure) SetBadHostHandler(handler http.Handler) {
 // Handler implements the http.HandlerFunc for integration with the standard net/http lib.
 func (s *Secure) Handler(h http.Handler) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if s.opt.nonceEnabled {
+			ctx := context.WithValue(r.Context(), nonceKey, randNonce(rand.NewSource(time.Now().UnixNano()), s.opt.CSPNonceByteAmount))
+			*r = *r.WithContext(ctx)
```

Comment:
I think just a `r = r.WithContext(ctx)` is what you want here. WithContext already returns a shallow copy, doing another shallow copy here seems unnecessary.

Reply:
That would work except in this case
```go
func handler(w http.ResponseWriter, r *http.Request) {
    next.ServrHTTP(w, r) // Nonce happens here
    secure.Nonce(r) // returns ""
}
```

Reply:
That's true, but I think that's actually what I'd expect.
Since this package is middleware, I'd expect it to have effect during the execution of the wrapped http.Handler, but not after that. It seems weird to me to have it "leak" permanent state into the request (especially because of the convention that the Request is basically immutable (hence the WithContext, rather than a SetContext())).

Reply:
What if someone needs a Footer to be sent?

Reply:
The Footer-sending code should also be "inside" the middleware.

In other words, I imagine the typical composition of handlers will be something like "secure(router(addFooter(userCode)))". So inside addFooter and userCode, all the necessary data will be present in the context; both the CSP nonce, and any router vars.

Indeed, its rather important that the "secure" middleware come first (or some equivalent of it), and not be in the middle of the middleware chain. For instance, if a server composed its middleware like this "middlewareA(secure(middlewareB(handler)))", then if middlewareA "returns early" before proceeding to the next stage in the chain (for example, if it performs request authorization, and the request is not allowed), then it does not get any of the security headers set by the secure middleware (nor for that matter would it get the CSP nonce, because this is before the Nonce function is invoked). Whereas, if you compose it like "secure(middlewareA(middlewareB(handler)))", then you do not have those problems. 

To confirm that my expectations are not crazy, I looked at the source code of two routers (gorilla/mux and "chi"), and they similarly only make route variables available "inside" their wrapped handlers. They do not write data into the request such that it leaks outside their respective scopes.